ksmt-core/src/main/kotlin/org/ksmt/cache/weak/ConcurrentWeakHashMapCache.kt

```diff
@@ -0,0 +1,321 @@
+                    scheduleCleanup()
+                }
+
+            } else {
+                // Acquire previously interned object
+                val currentKey = current.getKey()
+                if (currentKey != null) return currentKey
+
+                // Previously interned object was removed. Try cleanup and retry interning
+                scheduleCleanup()
+            }
+        }
+    }
+
+    private fun updateValueIfAlive(node: KeyRefNode<K, V>, value: V): V? = synchronized(node) {
+        if (!node.isAlive()) return null
+        val oldValue = node.getValue()
+        node.setValue(value)
+        oldValue
+    }
+
+    private fun notifyRemove(key: K?, value: V) {
+        removeHandler?.onRemove(key, value)
+    }
+
+    private fun afterRead() {
+        runCleanupIfRequired()
+    }
+
+    private fun afterWrite() {
+        if (modificationsSinceLastCleanup.incrementAndGet() >= MODIFICATIONS_TO_CLEANUP) {
+            modificationsSinceLastCleanup.set(0)
+            scheduleCleanup()
+        } else {
+            runCleanupIfRequired()
+        }
+    }
+
+    private fun runCleanupIfRequired() {
+        if (cleanupStatus.get() == REQUIRED) {
+            runCleanup()
+        }
+    }
+
+    private fun scheduleCleanup() {
+        while (true) {
+            when (cleanupStatus.get()) {
+                // No ongoing cleanup -> schedule and run
+                IDLE -> {
+                    cleanupStatus.compareAndSet(IDLE, REQUIRED)
+                    runCleanup()
+                    return
+                }
+                // No ongoing cleanup and cleanup is already scheduled -> run
+                REQUIRED -> {
+                    runCleanup()
+                    return
+                }
+                // Cleanup is running. Try to reschedule cleanup after completion.
+                PROCESSING_TO_IDLE -> {
+                    if (cleanupStatus.compareAndSet(PROCESSING_TO_IDLE, PROCESSING_TO_REQUIRED)) {
+                        return
+                    }
+                    // Cleanup status changed. Retry
+                }
+                // Cleanup is running and will be rescheduled right after completion.
+                PROCESSING_TO_REQUIRED -> {
+                    return
+                }
+            }
+        }
+    }
+
+    private inline fun ifNotProcessing(body: () -> Unit) {
+        val status = cleanupStatus.get()
+        if (status == PROCESSING_TO_IDLE || status == PROCESSING_TO_REQUIRED) {
+            return
+        }
+        body()
+    }
+
+    private fun runCleanup() = ifNotProcessing {
+        if (cleanupLock.tryLock()) {
+            try {
+                ifNotProcessing {
+                    cleanup()
+                }
+            } finally {
+                cleanupLock.unlock()
+            }
+        }
+    }
+
+    private fun cleanup() {
+        cleanupStatus.set(PROCESSING_TO_IDLE)
```

Comment:
Why we do not need a CAS here? We call this function only under a lock, but someone could already change the unlocked `cleanupStatus`, couldn't he?